Source/Urho3D/Actions/ActionBuilder.h

```diff
@@ -0,0 +1,189 @@
+
+    /// Continue with Blink action.
+    ActionBuilder Blink(float duration, unsigned numOfBlinks);
+
+    /// Continue with AttributeBlink action.
+    ActionBuilder Blink(float duration, unsigned numOfBlinks, ea::string_view attributeName);
+
+    /// Continue with AttributeTo action.
+    ActionBuilder AttributeTo(float duration, ea::string_view attributeName, const Variant& to);
+
+    /// Continue with AttributeFromTo action.
+    ActionBuilder AttributeFromTo(
+        float duration, ea::string_view attributeName, const Variant& from, const Variant& to);
+
+    /// Continue with ShaderParameterTo action.
+    ActionBuilder ShaderParameterTo(float duration, ea::string_view parameter, const Variant& to);
+
+    /// Continue with ShaderParameterFromTo action.
+    ActionBuilder ShaderParameterFromTo(
+        float duration, ea::string_view parameter, const Variant& from, const Variant& to);
+
+    /// Continue with SendEvent action.
+    ActionBuilder SendEvent(ea::string_view eventType, const StringVariantMap& data);
+
+    /// Continue with CallFunc action.
+    ActionBuilder CallFunc(Actions::ActionCallHandler* handler);
+
+    /// Continue with CallFunc action.
+    template <typename T>
+    ActionBuilder CallFunc(
+        T* receiver, typename Actions::ActionCallHandlerImpl<T>::HandlerFunctionPtr func, void* userData = nullptr)
+    {
+        return CallFunc(new Actions::ActionCallHandlerImpl<T>(receiver, func, userData));
+    }
+
+    /// Combine with BackIn action.
+    ActionBuilder BackIn();
+
+    /// Combine with BackOut action.
+    ActionBuilder BackOut();
+
+    /// Combine with BackInOut action.
+    ActionBuilder BackInOut();
+
+    /// Combine with BounceOut action.
+    ActionBuilder BounceOut();
+
+    /// Combine with BounceIn action.
+    ActionBuilder BounceIn();
+
+    /// Combine with BounceInOut action.
+    ActionBuilder BounceInOut();
+
+    /// Combine with SineOut action.
+    ActionBuilder SineOut();
+
+    /// Combine with SineIn action.
+    ActionBuilder SineIn();
+
+    /// Combine with SineInOut action.
+    ActionBuilder SineInOut();
+
+    /// Combine with ExponentialOut action.
+    ActionBuilder ExponentialOut();
+
+    /// Combine with ExponentialIn action.
+    ActionBuilder ExponentialIn();
+
+    /// Combine with ExponentialInOut action.
+    ActionBuilder ExponentialInOut();
+
+    /// Combine with ElasticIn action.
+    ActionBuilder ElasticIn(float period = 0.3f);
+
+    /// Combine with ElasticOut action.
+    ActionBuilder ElasticOut(float period = 0.3f);
+
+    /// Combine with ElasticInOut action.
+    ActionBuilder ElasticInOut(float period = 0.3f);
+
+    /// Combine with RemoveSelf action.
+    ActionBuilder RemoveSelf();
+
+    /// Combine with DelayTime action.
+    ActionBuilder DelayTime(float duration);
+
+    /// Repeat current action.
+    ActionBuilder Repeat(unsigned times);
+
+    /// Repeat current action forever (until canceled).
+    ActionBuilder RepeatForever();
```

Comment:
This and similar APIs return a copy of `ActionBuilder`, so object is copied for each added step. Would it be possible to use a single `ActionBuilder` instance, and return a reference here? If design makes it very hard to do, this is not the worst thing, actions will be built infrequently after all. Also, if this is the case on C# side (returning object copies), this is a bit of pressure on GC. Again, probably not something that would block merge..

Reply:
ActionBuilder is a very lightweight object. It could be a struct on c# side too, then it won't put any pressure on GC. I don't know how to map a c++ class on c# struct... Maybe I can copy something from Vector3.

Reply:
I can make ActionBuilder lighter. About of size of one pointer. There are several ways of doing it:

- I can split the ActionBuilder class into 2. One going to contain an action, another - context. But this would lead to code duplication.
- I can use ea::variant to store either context pointer or action shared pointer. This makes code overcomplicated.
- I can store SharedPtr to RefCounted but I don't know a cheap way of checking what object is in the SharedPtr as Context isn't Object for obvious reason.

I think returning a copy is a common pattern for Builder Pattern in c++. If you have a better solution - I can try it.

Reply:
All this sounds like really not worth the effort. None of this code will be a bottleneck i imagine. If we can't avoid copying without resorting to weird things that will make everything more confusing then so be it